File: workflow/scripts/seurat/01_qc.R
#!/usr/bin/env Rscript

# Quality Control Analysis with Seurat
# =====================================

suppressPackageStartupMessages({
  library(Seurat)
  library(ggplot2)
  library(patchwork)
  library(dplyr)
})

# Parse snakemake inputs
samples_file <- snakemake@params[["samples_file"]]
min_genes <- snakemake@params[["min_genes"]]
max_genes <- snakemake@params[["max_genes"]]
max_mito <- snakemake@params[["max_mito"]]
min_cells <- snakemake@params[["min_cells"]]

# Output files
qc_metrics_file <- snakemake@output[["metrics"]]
qc_plots_file <- snakemake@output[["plots"]]
seurat_obj_file <- snakemake@output[["seurat_obj"]]

# Load sample metadata
samples_df <- read.table(samples_file, header = TRUE, sep = "\t")

# Initialize list to store Seurat objects
seurat_list <- list()

# Load data for each sample
for (i in 1:nrow(samples_df)) {
  sample_id <- samples_df$sample_id[i]
  condition <- samples_df$condition[i]
  batch <- samples_df$batch[i]
  
  cat(sprintf("Loading sample: %s\n", sample_id))
  
  # Load 10x data
  h5_file <- sprintf("results/cellranger/%s/outs/filtered_feature_bc_matrix.h5", sample_id)
  
  if (file.exists(h5_file)) {
    data <- Read10X_h5(h5_file)
  } else {
    # Alternative: load from matrix directory
    data_dir <- sprintf("results/cellranger/%s/outs/filtered_feature_bc_matrix", sample_id)
    data <- Read10X(data_dir)
  }
  
  # Create Seurat object
  seurat_obj <- CreateSeuratObject(
    counts = data,
    project = sample_id,
    min.cells = min_cells,
    min.features = min_genes
  )
  
  # Add metadata
  seurat_obj$sample <- sample_id
  seurat_obj$condition <- condition
  seurat_obj$batch <- batch
  
  # Calculate mitochondrial percentage
  seurat_obj[["percent.mt"]] <- PercentageFeatureSet(seurat_obj, pattern = "^MT-")
  
  # Calculate ribosomal percentage
  seurat_obj[["percent.ribo"]] <- PercentageFeatureSet(seurat_obj, pattern = "^RP[SL]")
  
  seurat_list[[sample_id]] <- seurat_obj
}

# Merge all samples
if (length(seurat_list) > 1) {
  seurat_combined <- merge(
    seurat_list[[1]],
    y = seurat_list[2:length(seurat_list)],
    add.cell.ids = names(seurat_list),
    project = "scRNAseq"
  )
} else {
  seurat_combined <- seurat_list[[1]]
}

# Extract QC metrics before filtering
qc_metrics <- data.frame(
  cell_id = colnames(seurat_combined),
  sample = seurat_combined$sample,
  condition = seurat_combined$condition,
  batch = seurat_combined$batch,
  nCount_RNA = seurat_combined$nCount_RNA,
  nFeature_RNA = seurat_combined$nFeature_RNA,
  percent_mt = seurat_combined$percent.mt,
  percent_ribo = seurat_combined$percent.ribo
)

# Save pre-filtering metrics
write.csv(qc_metrics, qc_metrics_file, row.names = FALSE)

# Generate QC plots
pdf(qc_plots_file, width = 12, height = 10)

# Violin plots
p1 <- VlnPlot(seurat_combined, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"),
              ncol = 3, pt.size = 0.1)
print(p1)

# Scatter plots
p2 <- FeatureScatter(seurat_combined, feature1 = "nCount_RNA", feature2 = "percent.mt")
p3 <- FeatureScatter(seurat_combined, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
print(p2 + p3)

# Histograms
p4 <- ggplot(qc_metrics, aes(x = nFeature_RNA)) +
  geom_histogram(bins = 50, fill = "steelblue", color = "black") +
  geom_vline(xintercept = c(min_genes, max_genes), linetype = "dashed", color = "red") +
  theme_classic() +
  labs(title = "Distribution of Genes per Cell", x = "Number of Genes", y = "Count")

p5 <- ggplot(qc_metrics, aes(x = percent_mt)) +
  geom_histogram(bins = 50, fill = "coral", color = "black") +
  geom_vline(xintercept = max_mito, linetype = "dashed", color = "red") +
  theme_classic() +
  labs(title = "Distribution of Mitochondrial Content", x = "Percent Mitochondrial", y = "Count")

print(p4 / p5)

dev.off()

# Apply QC filters
seurat_filtered <- subset(
  seurat_combined,
  subset = nFeature_RNA > min_genes &
           nFeature_RNA < max_genes &
           percent.mt < max_mito
)

# Print filtering summary
cat("\n=== QC Filtering Summary ===\n")
cat(sprintf("Cells before filtering: %d\n", ncol(seurat_combined)))
cat(sprintf("Cells after filtering: %d\n", ncol(seurat_filtered)))
cat(sprintf("Cells removed: %d (%.1f%%)\n",
            ncol(seurat_combined) - ncol(seurat_filtered),
            100 * (ncol(seurat_combined) - ncol(seurat_filtered)) / ncol(seurat_combined)))

# Save filtered Seurat object
saveRDS(seurat_filtered, file = seurat_obj_file)

cat("\nQC analysis complete!\n")

File: workflow/scripts/scanpy/01_qc.py
#!/usr/bin/env python3

"""
Quality Control Analysis with Scanpy
=====================================
"""

import scanpy as sc
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from pathlib import Path

# Set plotting parameters
sc.settings.verbosity = 3
sc.settings.set_figure_params(dpi=300, facecolor='white', frameon=False)

# Parse snakemake inputs
samples_file = snakemake.params["samples_file"]
min_genes = snakemake.params["min_genes"]
max_genes = snakemake.params["max_genes"]
max_mito = snakemake.params["max_mito"]
min_cells = snakemake.params["min_cells"]

# Output files
qc_metrics_file = snakemake.output["metrics"]
qc_plots_file = snakemake.output["plots"]
adata_file = snakemake.output["adata"]

# Load sample metadata
samples_df = pd.read_csv(samples_file, sep='\t')

# Initialize list to store AnnData objects
adata_list = []

# Load data for each sample
for idx, row in samples_df.iterrows():
    sample_id = row['sample_id']
    condition = row['condition']
    batch = row['batch']
    
    print(f"Loading sample: {sample_id}")
    
    # Load 10x data
    h5_file = f"results/cellranger/{sample_id}/outs/filtered_feature_bc_matrix.h5"
    
    if Path(h5_file).exists():
        adata = sc.read_10x_h5(h5_file)
    else:
        # Alternative: load from matrix directory
        data_dir = f"results/cellranger/{sample_id}/outs/filtered_feature_bc_matrix"
        adata = sc.read_10x_mtx(data_dir, var_names='gene_symbols', cache=True)
    
    # Make variable names unique
    adata.var_names_make_unique()
    
    # Add metadata
    adata.obs['sample'] = sample_id
    adata.obs['condition'] = condition
    adata.obs['batch'] = batch
    
    adata_list.append(adata)

# Concatenate all samples
adata = sc.concat(adata_list, label="sample", keys=[row['sample_id'] for _, row in samples_df.iterrows()])

# Calculate QC metrics
adata.var['mt'] = adata.var_names.str.startswith('MT-')
adata.var['ribo'] = adata.var_names.str.match('^RP[SL]')

sc.pp.calculate_qc_metrics(
    adata,
    qc_vars=['mt', 'ribo'],
    percent_top=None,
    log1p=False,
    inplace=True
)

# Extract QC metrics before filtering
qc_metrics = pd.DataFrame({
    'cell_id': adata.obs_names,
    'sample': adata.obs['sample'],
    'condition': adata.obs['condition'],
    'batch': adata.obs['batch'],
    'n_genes_by_counts': adata.obs['n_genes_by_counts'],
    'total_counts': adata.obs['total_counts'],
    'pct_counts_mt': adata.obs['pct_counts_mt'],
    'pct_counts_ribo': adata.obs['pct_counts_ribo']
})

# Save pre-filtering metrics
qc_metrics.to_csv(qc_metrics_file, index=False)

# Generate QC plots
fig, axes = plt.subplots(2, 3, figsize=(15, 10))

# Violin plots
sc.pl.violin(adata, ['n_genes_by_counts', 'total_counts', 'pct_counts_mt'],
             jitter=0.4, multi_panel=True, ax=axes[0], show=False)

# Scatter plots
sc.pl.scatter(adata, x='total_counts', y='pct_counts_mt', ax=axes[1, 0], show=False)
sc.pl.scatter(adata, x='total_counts', y='n_genes_by_counts', ax=axes[1, 1], show=False)

# Histogram
axes[1, 2].hist(adata.obs['n_genes_by_counts'], bins=50, color='steelblue', edgecolor='black')
axes[1, 2].axvline(min_genes, color='red', linestyle='--', label=f'Min: {min_genes}')
axes[1, 2].axvline(max_genes, color='red', linestyle='--', label=f'Max: {max_genes}')
axes[1, 2].set_xlabel('Number of Genes')
axes[1, 2].set_ylabel('Count')
axes[1, 2].set_title('Distribution of Genes per Cell')
axes[1, 2].legend()

plt.tight_layout()
plt.savefig(qc_plots_file)
plt.close()

# Apply QC filters
n_cells_before = adata.n_obs

sc.pp.filter_cells(adata, min_genes=min_genes)
sc.pp.filter_genes(adata, min_cells=min_cells)

# Additional filters
adata = adata[adata.obs.n_genes_by_counts < max_genes, :]
adata = adata[adata.obs.pct_counts_mt < max_mito, :]

n_cells_after = adata.n_obs

# Print filtering summary
print("\n=== QC Filtering Summary ===")
print(f"Cells before filtering: {n_cells_before}")
print(f"Cells after filtering: {n_cells_after}")
print(f"Cells removed: {n_cells_before - n_cells_after} ({100 * (n_cells_before - n_cells_after) / n_cells_before:.1f}%)")

# Save filtered AnnData object
adata.write(adata_file)

print("\nQC analysis complete!")
